src/coreclr/src/pal/src/loader/module.cpp

```diff
@@ -1762,7 +1784,17 @@ MODSTRUCT *LOADGetPalLibrary()
             }
         }
 
-        pal_module = (MODSTRUCT *)LOADLoadLibrary(info.dli_fname, FALSE);
+        int len = strlen(g_szCoreCLRPath);
+        int suffixLen = strlen(PAL_SHLIB_SUFFIX);
+        if (len > suffixLen && strcmp(g_szCoreCLRPath + len - suffixLen, PAL_SHLIB_SUFFIX) != 0)
```

Comment:
This check is not particularly robust. If I rename my single file app binary to something.so, it will stop working.

Should we set it `g_running_in_exe` explicitly for super host only?

Reply:
I am going to try the following solution:
- coreclr DllMain sets some internal flag like `g_dllmain_has_run`
- if `coreclr_initialize` sees the flag not set, it runs DllMain first of all things. 

This is to address the concerns about fragility of the contract. Also should be less platform specific.

If that works, I can just assume that we are "exe-initializing"  when `coreclr_initialize` needs to run DllMain

Reply:
This alone won't be sufficient. DllMain is used both as a process initialization callback and as thread initialization/shutdown callback. You will also need to do something about the thread initialization/shutdown part.

Reply:
Right. I was mostly thinking about process attach.  

I do not want to make this more complex than necessary but it feels like the mechanism for this should be 
- something that works on Windows too, and 
- preferably should be something that will allow having more than one embedded library like coreclr. Although this is not a requirement right now and coreclr is indeed quite special.

Reply:
For the PAL_INITIALIZE_xxx flag - is the suggested way of implementing it is the same way as  BUNDLE_PROBE and other properties?

Reply:
Actually, I have just realized we can do without an extra flag. The coreclr_initialize passes the exePath that is required to be the absolute path to the PAL. 
It seems that we can just store that exePath in PAL_InitializeCoreCLR in some global variable and at the place whhere we compute the g_running_in_exe, we would just compare g_szCoreCLRPath to that exe path and if it matches, we have coreclr linked into the superhost.

Reply:
I think flag would be better. Comparing paths for equality is a factory for corner case bugs.

Reply:
I was thinking about the Linux case where I believe the paths should always match character to character, but I am not sure about other systems, so I guess you are right.
@VSadov then I guess the BUNDLE_PROBE way would do.